Objects/typeobject.c

```diff
@@ -2938,43 +2942,21 @@ PyType_GetSlot(PyTypeObject *type, int slot)
     return  *(void**)(((char*)type) + slotoffsets[slot]);
 }
 
-/* Internal API to look for a name through the MRO.
-   This returns a borrowed reference, and doesn't set an exception! */
-PyObject *
-_PyType_Lookup(PyTypeObject *type, PyObject *name)
+/* Internal API to look for a name through the MRO, bypassing the method cache.
+   This returns a borrowed reference, and might set an exception! */
+static PyObject *
+_PyType_LookupUncached(PyTypeObject *type, PyObject *name, int *error)
 {
     Py_ssize_t i, n;
     PyObject *mro, *res, *base, *dict;
-    unsigned int h;
-
-    if (MCACHE_CACHEABLE_NAME(name) &&
-        PyType_HasFeature(type, Py_TPFLAGS_VALID_VERSION_TAG)) {
-        /* fast path */
-        h = MCACHE_HASH_METHOD(type, name);
-        if (method_cache[h].version == type->tp_version_tag &&
-            method_cache[h].name == name) {
-#if MCACHE_STATS
-            method_cache_hits++;
-#endif
-            return method_cache[h].value;
-        }
-    }
-
+    Py_hash_t hash;
+    *error = 1;
```

Comment:
Alternatively you could return a special value for signalling error. E.g. (PyObject *)(-1). Test what is faster.

Reply:
I can set `error` to `-1` on exceptions. That will distinguish all three cases: ok, error with exception, error without exception. Only `-1` will need a call to `PyErr_Clear()` then.

Reply:
I dislike things like `(PyObject *)(-1)` because I wouldn't want to assume that they are not valid pointer values on any system whatsoever.